lib/autoupdate/rolloutcontroller/reconciler.go

```diff
@@ -0,0 +1,222 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package rolloutcontroller
+
+import (
+	"context"
+	"github.com/gravitational/teleport/api/gen/proto/go/teleport/autoupdate/v1"
+	update "github.com/gravitational/teleport/api/types/autoupdate"
+	"github.com/gravitational/trace"
+	"log/slog"
+	"sync"
+	"time"
+)
+
+const (
+	reconciliationTimeout = 30 * time.Second
+	defaultConfigMode     = update.AgentsUpdateModeEnabled
+	defaultStrategy       = update.AgentsStrategyHaltOnError
+)
+
+// Reconciler reconciles the AutoUpdateAgentRollout singleton based on the content of the AutoUpdateVersion and
+// AutoUpdateConfig singletons. This reconciler is not based on the services.GenericReconciler because:
+// - we reconcile 2 resources with one
+// - both input and output are singletons, we don't need the multi resource logic nor stream/paginated APIs
+type Reconciler struct {
+	clt Client
+	log *slog.Logger
+
+	// lock ensures we only run one reconciliation at a time
+	lock sync.Mutex
+}
+
+// Reconcile the AutoUpdateAgentRollout singleton. The reconciliation can fail because of a conflict (multiple auths
+// are racing), in this case we retry the reconciliation immediately.
+func (r Reconciler) Reconcile(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, reconciliationTimeout)
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			err := r.tryReconcile(ctx)
+			switch {
+			case err == nil:
+				return nil
+			case trace.IsCompareFailed(err), trace.IsNotFound(err):
+				// The resource changed since we last saw it
+				// We must have raced against another auth
+				// Let's retry the reconciliation
+				// TODO: add a backoff to avoid a tight or infinite loop?
```

Comment:
I would definitely add a backoff, even a small one, or maybe an iteration limit for a single `Reconcile` call?